javatests/arcs/core/data/TypeUtilsTest.kt

```diff
@@ -178,6 +223,100 @@ class TypeUtilsTest {
     )
   }
 
+  @Test
+  fun constructTypeScope_emptyMapScope() {
+    val typeMap = emptyMap<String, Type>()
+
+    val scope = constructTypeScope(typeMap)
+    assertThat(scope.scopeName).isEqualTo("root")
+    assertThat(scope.properties()).isEqualTo(emptySet<String>())
+  }
+
+  @Test
+  fun constructTypeScope_mapScopeWithSingletonAndCollection() {
+    SchemaRegistry.register(REFERENCED_DUMMY_SCHEMA)
+    SchemaRegistry.register(DUMMY_SCHEMA)
+
+    val ent = EntityType(DUMMY_SCHEMA)
+    val scope = constructTypeScope(
+      mapOf(
+        testField to SingletonType(ent),
+        testField2 to CollectionType(ent)
+      )
+    )
+    assertThat(scope.scopeName).isEqualTo("root")
+    assertThat(scope.properties()).isEqualTo(setOf<String>(testField, testField2))
+
+    val dummyType = mapTypeToInferredType(ent)
+    assertThat(scope.lookup<InferredType>(testField)).isEqualTo(dummyType)
+    assertThat(scope.lookup<InferredType>(testField2)).isEqualTo(InferredType.SeqType(dummyType))
+  }
+
+  @Test
+  fun typeCheck_fieldExpressionHasTypeNumber() {
+    SchemaRegistry.register(REFERENCED_DUMMY_SCHEMA)
+    SchemaRegistry.register(DUMMY_SCHEMA)
+
+    val expr = Expression.FieldExpression<Any>(null, testField, false)
+
+    val ent = EntityType(DUMMY_SCHEMA)
+
+    assertThat(typeCheck(
+      "testConnectionName",
+      expr,
+      constructTypeScope(mapOf(testField to SingletonType(ent))),
+      SingletonType(ent)
+    )).isEqualTo(emptyList<String>()) // These are the warnings
+  }
+
+  @Test
+  fun typeCheck_addExpressionHasTypeNumber() {
+    SchemaRegistry.register(REFERENCED_DUMMY_SCHEMA)
+    SchemaRegistry.register(DUMMY_SCHEMA)
+
```

Comment:
With this being a unit test, I think we should probably avoid the extra complexity. Happy to take your advice on it though.

Reply:
Also, doesn't this overlap with the TypeCheckerTest? There's tests in there for binary expressions, and field expressions? Did you just choose this one because you wanted a non-trivial example?

Reply:
Yes. Just wanted to ensure that there was some basic functionality tested in a unit test so that (if/when it were to fail) we'd know it wasn't the builders or the evaluator particle that was broken, but the typeCheck function.